libs/cmdio/render.go

```diff
@@ -46,8 +50,117 @@ func Heredoc(tmpl string) (trimmed string) {
 	return strings.TrimSpace(trimmed)
 }
 
-func renderJson(w io.Writer, v any) error {
-	pretty, err := fancyJSON(v)
+// writeFlusher represents a buffered writer that can be flushed. This is useful when
+// buffering writing a large number of resources (such as during a list API).
+type writeFlusher interface {
+	io.Writer
+	Flush() error
+}
+
+type jsonRenderer interface {
+	// Render an object as JSON to the provided writeFlusher.
+	renderJson(context.Context, writeFlusher) error
+}
+
+type textRenderer interface {
+	// Render an object as text to the provided writeFlusher.
+	renderText(context.Context, writeFlusher) error
+}
+
+type templateRenderer interface {
+	// Render an object using the provided template and write to the provided writeFlusher.
+	renderTemplate(context.Context, *template.Template, writeFlusher) error
+}
+
+type readerRenderer struct {
+	reader io.Reader
+}
+
+func (r readerRenderer) renderText(_ context.Context, w writeFlusher) error {
+	_, err := io.Copy(w, r.reader)
+	if err != nil {
+		return err
+	}
+	return w.Flush()
+}
+
+type iteratorRenderer struct {
+	t          reflectIterator
+	bufferSize int
+}
+
+func (ir iteratorRenderer) getBufferSize() int {
+	if ir.bufferSize == 0 {
+		return 100
+	}
+	return ir.bufferSize
+}
+
+func (ir iteratorRenderer) renderJson(ctx context.Context, w writeFlusher) error {
+	// Iterators are always rendered as a list of resources in JSON.
+	_, err := w.Write([]byte("[\n  "))
+	if err != nil {
+		return err
+	}
+	for i := 0; ir.t.HasNext(ctx); i++ {
+		if i != 0 {
+			_, err = w.Write([]byte(",\n  "))
+			if err != nil {
+				return err
+			}
+		}
+		n, err := ir.t.Next(ctx)
+		if err != nil {
+			return err
+		}
+		res, err := json.MarshalIndent(n, "  ", "  ")
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(res)
+		if err != nil {
+			return err
+		}
+		if (i+1)%ir.getBufferSize() == 0 {
+			err = w.Flush()
+			if err != nil {
+				return err
+			}
+		}
+	}
+	_, err = w.Write([]byte("\n]\n"))
+	if err != nil {
+		return err
+	}
+	return w.Flush()
+}
+
+func (ir iteratorRenderer) renderTemplate(ctx context.Context, t *template.Template, w writeFlusher) error {
+	for i := 0; ir.t.HasNext(ctx); i++ {
+		n, err := ir.t.Next(ctx)
+		if err != nil {
+			return err
+		}
+		err = t.Execute(w, []any{n})
```

Comment:
This means each row is rendered independently and column sizing doesn't take into account multiple rows.

Reply:
Turns out `w` is a `tabwriter`, not the buffered flusher defined below.

Looking at the implementation, the `w` argument can be concretely typed as the `tabwriter`.

Reply:
I thought it would be "nicer" if all of the formatting interfaces used the same type, but if it obscures the behavior too much, happy to undo this.